core/Plugin/Controller.php

```diff
@@ -953,6 +955,41 @@ public static function getPrettyDate($date, $period)
         return self::getCalendarPrettyDate(Period\Factory::build($period, Date::factory($date)));
     }
 
+    /**
+     * Return the "long" pretty date representation.
+     *
+     * Uses the **date** and **period** parameters from the query string.
+     *
+     * @return string Long pretty date
+     */
+    protected function getLongPrettyDate()
+    {
+        $siteTimezone = $this->site->getTimezone();
+
+        $rawDate = Common::getRequestVar('date');
+        Period::checkDateFormat($rawDate);
+
+        $periodStr = Common::getRequestVar('period');
+
+        if ($periodStr != 'range') {
+            $date      = Date::factory($this->strDate);
+            $period    = Period\Factory::build($periodStr, $date);
+        } else {
+            $period = new Range($periodStr, $rawDate, $siteTimezone);
+        }
+
+        $dateStart = $period->getDateStart();
+        $dateEnd   = $period->getDateEnd();
+
+        if ($periodStr == 'range') {
+            // make sure we actually display the correct calendar pretty date
+            $newRawDate = $dateStart->toString() . ',' . $dateEnd->toString();
+            $period = new Range($periodStr, $newRawDate, $siteTimezone);
+        }
+
+        return $period->getLocalizedLongString();
+    }
```

Comment:
I think if we can avoid having this logic in two places, it would be better. From looking at the `Controller::setGeneralVariablesView()` method, it looks like the reason it's difficult to pull this out, is because each variable (or nearly each) is set as a property on the view.

Maybe we can create a `getCurrentPeriodInfo()` method (w/ or w/o a better name) that returns an array with all the information needed in `setGeneralVariablesView()`. In `setGeneralVariablesView()`, you'd set the result as properties on `$view`, and in `CoreHome:: getDateString()` you'd just use the localized long string. Can you do something like that here?

Might look something like:

```
class Controller
{
    protected function getCurrentPeriodInfo()
    {
        $dateStart = // ...
        $dateEnd = // ...
        // ...
        $prettyDateLong = // ...

        return [
            'dateStart' => $dateStart,
            'dateEnd' => $dateEnd,
            // ...
            'prettyDateLong' => $prettyDateLong,
        ];
    }

    protected function setGeneralVariablesView($view)
    {
        // ...

        $dateInfo = $this->getCurrentPeriodInfo();
        foreach ($dateInfo as $name => $value) {
            $view->$name = $value;
        }

        // ...
    }
}
```

Reply:
Sounds good to me 👍 I used the name `getCurrentDatePeriodVars()` (can change if advised). 

I struggled a bit to extract the side-effects that are tangled up with this "date and period" logic: specifically, the calls to `Period::checkDateFormat($rawDate)` which throws if date is invalid, and `$this->setDate($validDate)`, which modifies a property on `$this`. 

I decided to leave these side-effects in the `getCurrentDatePeriodVars()` method. For the former, this validation seems important. For the latter, there would be a fair amount of logic left duplicated. Let me know if you have any input on this!

Reply:
I see what you mean about `setDate()`, those side effects could result in confusion in the future. `setDate()` sets the `$this->strDate` & `$this->date` properties on the controller, but looks like they aren't used in `getCurrentDatePeriodVars()`, so we could potentially move the `setDate()` call to after `$dateVars = $this->getCurrentDatePeriodVars();`.

Something like:

```
// in getCurrentDatePeriodVars() return the $validDate variable too...
$validDate = null;
if ($periodStr != 'range') {
    // ...

    $validDate = ...
} else {
    // ...
}

// ...

return [
    // ...
    'validDate' => $validDate,
];
```

then in `setGeneralVariablesView()`:

```
$dateVars = $this->getCurrentDatePeriodVars();
if (Date::factory($this->strDate)->toString() != $validDate->toString()) {
    // ... comment ...
    $this->setDate($validDate);
}
```

then the side effect will be limited to the setGeneralVariablesView() method. Think it'll work?

Reply:
Yep, I think that will work just fine!